File: 20221031_Monday/20221031.go
// Leetcode: 766. Toeplitz Matrix.
// https://leetcode.com/problems/toeplitz-matrix/
// = = = = = = = = = = = = = =
// Accpeted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 20 ms, faster than 44.12% of Go online submissions for Toeplitz Matrix.
// Memory Usage: 4.3 MB, less than 67.65% of Go online submissions for Toeplitz Matrix.
// 2022.10.31 Daily Challenge.
var r int
var c int
func isToeplitzMatrix(matrix [][]int) bool {
  r,c = len(matrix), len(matrix[0])
  start_j := 0
  for start_i := 0; start_i < r ; start_i += 1 {
    if ! check_diagonal(start_i, start_j, matrix) {
      return false
    }
  }
  start_i := 0
  for start_j := 0; start_j < c; start_j += 1 {
    if ! check_diagonal(start_i, start_j, matrix) {
      return false
    }
  }
  return true
}

func check_diagonal(start_i, start_j int, matrix [][]int) bool{
  prv := -1
  i,j := start_i, start_j
  for ; i >= 0 && i < r && j >= 0 && j < c; {
    if -1 == prv {
      prv = matrix[i][j]
    }
    if prv != matrix[i][j] {
      return false
    }
    prv = matrix[i][j]
    i += 1
    j += 1
  }
  return true
}

File: 20221031_Monday/20221031.rb
# Leetcode: 766. Toeplitz Matrix.
# https://leetcode.com/problems/toeplitz-matrix/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 88 ms, faster than 81.82% of Ruby online submissions for Toeplitz Matrix.
# Memory Usage: 211.5 MB, less than 18.18% of Ruby online submissions for Toeplitz Matrix.
# 2022.10.31 Daily Challenge.
# @param {Integer[][]} matrix
# @return {Boolean}
def is_toeplitz_matrix(matrix)
  rows, cols = matrix.size, matrix[0].size
  j_start = 0
  (0...rows).each do |i_start|
    previous = nil 
    i, j = i_start, j_start
    while i.between?(0,rows-1) && j.between?(0,cols-1)
      previous = matrix[i][j] if previous.nil?
      return false if previous != matrix[i][j]
      previous = matrix[i][j]
      i+=1
      j+=1
    end
  end
  i_start = 0
  (0...cols).each do |j_start|
    previous = nil
    i, j = i_start, j_start
    while i.between?(0,rows-1) && j.between?(0,cols-1)
      previous = matrix[i][j] if previous.nil?
      return false if previous != matrix[i][j]
      previous = matrix[i][j]
      i+=1
      j+=1
    end
  end
  return true
end

